src/backend/optimizer/plan/planshare.c

```diff
@@ -211,6 +211,14 @@ prepare_plan_for_sharing(PlannerInfo *root, Plan *common)
 		Material *m = make_material(common);
 		shared = (Plan *) m;
 
+		if (!shared->targetlist &&
+			 IsA(common, ModifyTable))
+		{
+			ModifyTable *mt = (ModifyTable *)common;
+			shared->targetlist = copyObject(linitial(mt->returningLists));
```

Comment:
may be add check?
```c
if (mt->returningLists)
```
or
```c
Assert(mt->returningLists);
```

Reply:
If we have modifying CTE, parser won't allow to write something without returning list.
```
explain (costs off)
with cte1 as (
  insert into t1 values (1,2)            
) select * from cte1 a join cte1 b using(c1);
ERROR:  WITH query "cte1" does not have a RETURNING clause
LINE 4: ) select * from cte1 a join cte1 b using(c1);
```
Moreover, if we don't have returningList this solution will be equivalent to the initial solution  with NIL targetList

Reply:
`linitial` and `copyObject` normally handle NULL values. so no verification is required.

Reply:
Fully expanded linitial is `((list_head(mt->returningLists))->data.ptr_value)`. The list_head returns NULL, when its argument is NULL. So if mt->returningLists is NULL, then segfault happens. I suggest to add a check